drivers/input/input_esp32_touch_sensor.c

```diff
@@ -0,0 +1,390 @@
+	case 2400000:
+		volt.refh = TOUCH_HVOLT_2V4;
+		break;
+	case 2500000:
+		volt.refh = TOUCH_HVOLT_2V5;
+		break;
+	case 2600000:
+		volt.refh = TOUCH_HVOLT_2V6;
+		break;
+	case 2700000:
+		volt.refh = TOUCH_HVOLT_2V7;
+		break;
+	}
+
+	switch (dev_cfg->lref_microvolt) {
+	case 500000:
+		volt.refl = TOUCH_LVOLT_0V5;
+		break;
+	case 600000:
+		volt.refl = TOUCH_LVOLT_0V6;
+		break;
+	case 700000:
+		volt.refl = TOUCH_LVOLT_0V7;
+		break;
+	case 800000:
+		volt.refl = TOUCH_LVOLT_0V8;
+		break;
+	}
+
+	switch (dev_cfg->lref_microvolt) {
+	case 0:
+		volt.atten = TOUCH_HVOLT_ATTEN_0V;
+		break;
+	case 500000:
+		volt.atten = TOUCH_HVOLT_ATTEN_0V5;
+		break;
+	case 1000000:
+		volt.atten = TOUCH_HVOLT_ATTEN_1V;
+		break;
+	case 1500000:
+		volt.atten = TOUCH_HVOLT_ATTEN_1V5;
+		break;
+	}
+
+#endif /* defined(CONFIG_SOC_SERIES_ESP32) */
+
+	touch_hal_init();
+#if defined(CONFIG_SOC_SERIES_ESP32)
+	touch_hal_set_voltage(&volt);
+	touch_hal_set_fsm_mode(TOUCH_FSM_MODE_TIMER);
+#endif /* defined(CONFIG_SOC_SERIES_ESP32) */
+
+	for (int i = 0; i < num_channels; i++) {
+		struct esp32_touch_sensor_channel_data
+				*channel_data = &dev_cfg->channel_data[i];
+		const struct esp32_touch_sensor_channel_config
+				*channel_cfg = &dev_cfg->channel_cfg[i];
+
+		if (!(channel_cfg->channel_num > 0 &&
+			channel_cfg->channel_num < SOC_TOUCH_SENSOR_NUM)) {
+			LOG_ERR("Touch %d configuration failed: "
+				"Touch channel error", i);
+			return -EINVAL;
+		}
+
+#if defined(CONFIG_SOC_SERIES_ESP32S2) || defined(CONFIG_SOC_SERIES_ESP32S3)
+		if (channel_cfg->channel_num == SOC_TOUCH_DENOISE_CHANNEL) {
+			LOG_ERR("Touch %d configuration failed: "
+				"TOUCH0 is internal denoise channel", i);
+			return -EINVAL;
+		}
+#endif  /* defined(CONFIG_SOC_SERIES_ESP32S2) || defined(CONFIG_SOC_SERIES_ESP32S3) */
+
+		gpio_num_t gpio_num = touch_sensor_channel_io_map[channel_cfg->channel_num];
+
+		rtc_gpio_init(gpio_num);
+		rtc_gpio_set_direction(gpio_num, RTC_GPIO_MODE_DISABLED);
+		rtc_gpio_pulldown_dis(gpio_num);
+		rtc_gpio_pullup_dis(gpio_num);
+
+		touch_hal_config(channel_cfg->channel_num);
+#if defined(CONFIG_SOC_SERIES_ESP32)
+		touch_hal_set_threshold(channel_cfg->channel_num, 0);
+		touch_hal_set_group_mask(BIT(channel_cfg->channel_num),
+					 BIT(channel_cfg->channel_num));
+#endif /* defined(CONFIG_SOC_SERIES_ESP32) */
+		touch_hal_set_channel_mask(BIT(channel_cfg->channel_num));
+
+		channel_data->status = 0;
+#if defined(CONFIG_SOC_SERIES_ESP32S2) || defined(CONFIG_SOC_SERIES_ESP32S3)
+		channel_data->last_status = 0;
+#endif /* defined(CONFIG_SOC_SERIES_ESP32S2) || defined(CONFIG_SOC_SERIES_ESP32S3) */
+		channel_data->dev = dev;
+		k_work_init_delayable(&channel_data->work,
+				      esp32_touch_sensor_change_deferred);
+	}
+
+#if defined(CONFIG_SOC_SERIES_ESP32)
+	for (int i = 0; i < num_channels; i++) {
+		const struct esp32_touch_sensor_channel_config
+				*channel_cfg = &dev_cfg->channel_cfg[i];
+
+		while (!touch_hal_meas_is_done()) {
+		}
```

Comment:
ouch, prob a good idea to bound this loop and return an initialization error rather than lock up the device at boot if this happens to never exit

Reply:
you're right. I'll implement a timeout mechanism.